cmd/contour/cli.go

```diff
@@ -118,21 +140,242 @@ type stream interface {
+	stream, err := envoy_service_endpoint_v3.NewEndpointDiscoveryServiceClient(c.dial()).DeltaEndpoints(context.Background())
+	if err != nil {
+		c.Log.WithError(err).Fatal("failed to fetch incremental stream of Endpoints")
+	}
+	return stream
+}
+
+// ListenerStream returns a stream of Listeners using the config in the Client.
+func (c *Client) DeltaListenerStream() envoy_service_listener_v3.ListenerDiscoveryService_DeltaListenersClient {
+	stream, err := envoy_service_listener_v3.NewListenerDiscoveryServiceClient(c.dial()).DeltaListeners(context.Background())
+	if err != nil {
+		c.Log.WithError(err).Fatal("failed to fetch incremental stream of Listeners")
+	}
+	return stream
+}
+
+// RouteStream returns a stream of Routes using the config in the Client.
+func (c *Client) DeltaRouteStream() envoy_service_route_v3.RouteDiscoveryService_DeltaRoutesClient {
+	stream, err := envoy_service_route_v3.NewRouteDiscoveryServiceClient(c.dial()).DeltaRoutes(context.Background())
+	if err != nil {
+		c.Log.WithError(err).Fatal("failed to fetch incremental stream of Routes")
+	}
+	return stream
+}
+
+type deltaStream interface {
+	Send(*envoy_discovery_v3.DeltaDiscoveryRequest) error
+	Recv() (*envoy_discovery_v3.DeltaDiscoveryResponse, error)
+}
+
+func watchDeltaStream(log *logrus.Logger, st deltaStream, typeURL string, resources []string, nack bool, nodeID string) {
+	m := protojson.MarshalOptions{
+		Multiline:     true,
+		Indent:        "  ",
+		UseProtoNames: true,
+	}
+
+	currentVersion := "0"
+
+	// Send the initial, non-ACK discovery request.
+	req := &envoy_discovery_v3.DeltaDiscoveryRequest{
+		TypeUrl:                typeURL,
+		ResourceNamesSubscribe: resources,
+		Node: &corev3.Node{
+			Id: nodeID,
+		},
+	}
+	log.WithField("currentVersion", currentVersion).Info("Sending incremental discover request")
+	fmt.Println(m.Format(req))
+	err := st.Send(req)
+	if err != nil {
+		log.WithError(err).Fatal("failed to send incremental Discover Request")
+	}
+
+	for {
+
+		// Wait until we receive a response to our request.
 		resp, err := st.Recv()
-		kingpin.FatalIfError(err, "failed to receive response for Discover Request")
-		err = m.Marshal(os.Stdout, resp)
-		kingpin.FatalIfError(err, "failed to marshal Discovery Response")
+		if err != nil {
+			log.WithError(err).Fatal("failed to receive response for incremental Discover Request")
+		}
+		log.WithField("currentVersion", currentVersion).
+			WithField("resp_system_version_info", resp.SystemVersionInfo).
+			WithField("nonce", resp.Nonce).
+			Info("Received Discovery Response")
+
+		fmt.Println(m.Format(resp))
+		if err != nil {
+			log.WithError(err).Fatal("failed to marshal incremental Discovery Response")
+		}
+
+		currentVersion = resp.SystemVersionInfo
+
+		if nack {
+			// We'll NACK the response we just got.
+			// The ResponseNonce field is what makes it an ACK,
+			// and the VersionInfo field must match the one in the response we
```

Comment:
similar thing here for the version info which is less important in the delta case and isn't part of the request, could probably remove some of the commentary about it here: https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#incremental-xds

Reply:
versions and response nonces will likely end up diverging since nonces are per-client req/resp and the snapshot/system version info is more based on config changes as well

Reply:
Yeah, fair enough, I'll remove the versioninfo stuff, but I think having some information about what makes it an ACK is useful.